issue-44801 EnsureCapacity Apis For List Stack Queue

lateapexearlyspeed opened the pull request:
Fix #44801 
Current status is to make sure there is a PR on progress, will notify when it is ready for review, thanks.

lateapexearlyspeed committed:
issue-44801 Initial commit: public List.EnsureCapacity.

eiriktsarpalis commented:
> Current status is to make sure there is a PR on progress, will notify when it is ready for review, thanks.

Thanks for taking this. Consider marking this as a draft PR while working on it.

eiriktsarpalis commented on src/libraries/System.Private.CoreLib/src/System/Collections/Generic/List.cs:
```diff
@@ -391,22 +391,26 @@ public void CopyTo(T[] array, int arrayIndex)
                 // Note that this check works even when _items.Length overflowed thanks to the (uint) cast
                 if ((uint)newCapacity > Array.MaxArrayLength) newCapacity = Array.MaxArrayLength;
-                if (newCapacity < min) newCapacity = min;
+                if (newCapacity < capacity) newCapacity = capacity;
```
The existing logic will pick `capacity` as its new size if `capacity > 2 * _items.Length`. Wondering if inserting something along the lines of
```csharp
while (newCapacity < capacity)
{
    newCapacity *= 2;
}
```
is preferable? @stephentoub thoughts?

lateapexearlyspeed replied:
Fixed.

lateapexearlyspeed committed:
Create Stack.EnsureCapacity().

eiriktsarpalis commented on src/libraries/System.Collections/src/System/Collections/Generic/Stack.cs:
```diff
@@ -288,6 +288,33 @@ private void PushWithResize(T item)
+                {
+                    newCapacity *= 2;
+                }
+                Array.Resize(ref _array, newCapacity);
```
Perhaps add a check for `Array.MaxArrayLength` like List does?

lateapexearlyspeed replied:
@eiriktsarpalis Fixed, please review and give feedback then I can do similar thing for List combining with List's comments, thanks :)

lateapexearlyspeed committed:
issue-44801 Fix comment for Stack.

eiriktsarpalis commented on src/libraries/System.Collections/src/System/Collections/Generic/Stack.cs:
```diff
@@ -31,6 +31,7 @@ public class Stack<T> : IEnumerable<T>,
         private int _version; // Used to keep enumerator in sync w/ collection. Do not rename (binary serialization)
 
         private const int DefaultCapacity = 4;
+        private const int MaxArrayLength = 0X7FEFFFFF; // This is the maximum array length value from Array.MaxArrayLength
```
This seems out of place but there is certainly precedent elsewhere, e.g. https://github.com/dotnet/runtime/blob/1d9e50cb4735df46d3de0cee5791e97295eaf588/src/libraries/Common/src/System/Collections/Generic/EnumerableHelpers.cs#L57

stephentoub replied:
This can go away once https://github.com/dotnet/runtime/pull/43301 lands.

lateapexearlyspeed replied:
Fixed by moving this temp value to nearby usage scope.

lateapexearlyspeed committed:
issue-44801 Fix comment: Move temp MaxArrayLength into near usage scope for Stack.

lateapexearlyspeed committed:
issue-44801 Fix comment for List.

lateapexearlyspeed commented on src/libraries/System.Private.CoreLib/src/System/Collections/Generic/List.cs:
```diff
@@ -391,22 +391,26 @@ public void CopyTo(T[] array, int arrayIndex)
                 if ((uint)newCapacity > Array.MaxArrayLength) newCapacity = Array.MaxArrayLength;
-                if (newCapacity < min) newCapacity = min;
+                if (newCapacity < capacity) newCapacity = capacity;
                 Capacity = newCapacity;
```
@eiriktsarpalis Fixed.

lateapexearlyspeed committed:
issue-44801 Create EnsureCapacity for Queue.

lateapexearlyspeed commented:
For Queue.EnsureCapacity(), I did not use private SetCapacity() as following check can be removed:

> _tail = (_size == capacity) ? 0 : _size;

Not sure if it deserved to 'inline' most of its implementation into Queue.EnsureCapacity() rather than calling SetCapacity().

lateapexearlyspeed committed:
issue-44801 Create tests for List.EnsureCapacity.

lateapexearlyspeed committed:
issue-44801 Create tests for Stack.EnsureCapacity.

lateapexearlyspeed committed:
issue-44801 Create tests for Queue.EnsureCapacity.

lateapexearlyspeed commented:
It is ready for review now.

lateapexearlyspeed committed:
issue-44801 Update version if updating internal buffer.

eiriktsarpalis commented:
It seems like many of the new tests are failing: https://dev.azure.com/dnceng/public/_build/results?buildId=975051&view=ms.vss-test-web.build-test-results-tab&runId=30602816&resultId=117111&paneView=debug

Could you take a look?

lateapexearlyspeed commented:
Should because of the last commit about version update which was the only one not run test in my local :( will fix later.

eiriktsarpalis commented on src/libraries/System.Collections/src/System/Collections/Generic/Queue.cs:
```diff
@@ -386,6 +386,64 @@ public void TrimExcess()
+                _array = newArray;
+                _head = 0;
+                _tail = _size;
+                _version++;
```
Should resize operations invalidate existing enumerators? Unless the collection is modified concurrently I don't see why it could cause issues cc @stephentoub

lateapexearlyspeed replied:
Checked code, maybe because EnsureCapacity() renew internal buffer instance so existing code treats it as version (data) update. MoveNext checked version (data) changed so it blocked current Enumerator.MoveNext ? Seems Dictionary.[EnsureCapacity ](https://github.com/dotnet/runtime/blob/bee5b795add1a3113925be13d25a0d419ff0d9c4/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/Dictionary.cs#L1037)and Dictionary.[Enumerator ](https://github.com/dotnet/runtime/blob/bee5b795add1a3113925be13d25a0d419ff0d9c4/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/Dictionary.cs#L1274)has same case and its test also verified it can [invalidate ](https://github.com/dotnet/runtime/blob/bee5b795add1a3113925be13d25a0d419ff0d9c4/src/libraries/System.Collections/tests/Generic/Dictionary/Dictionary.Generic.Tests.cs#L285)enumeration. So we should turn to check they do invalidate enumeration ?

lateapexearlyspeed committed:
issue-44801 Update test cases to verify invalidating enumeration (List, Stack, Queue).

lateapexearlyspeed commented:
@eiriktsarpalis Reopen cannot pass those failures, are they related to this PR ?

lateapexearlyspeed committed:
issue-44801 Improve code change for List:

1. Avoid incrementing version number twice in one Insert (or Add) method call;
2. Avoid more capacity check for Insert (or Add) method.

lateapexearlyspeed commented:
Hi @eiriktsarpalis, build passed after last new commit (although not sure if failure is related to it, maybe it just trigger refresh build)
Could you review this commit ? It is about some improvement as commit message describes.

eiriktsarpalis commented on src/libraries/System.Collections/src/System/Collections/Generic/Queue.cs:
```diff
@@ -386,6 +386,64 @@ public void TrimExcess()
+
+            if (_array.Length < capacity)
+            {
+                int newCapacity = (int)((long)(_array.Length == 0 ? 1 : _array.Length) * (long)GrowFactor / 100);
```
Why set the capacity to `1` if the array is empy? FWIW we should probably just replicate the logic as defined in the Enqueue method: https://github.com/dotnet/runtime/blob/cb5431d0c0790a69925145a15952a9601b9be102/src/libraries/System.Collections/src/System/Collections/Generic/Queue.cs#L186-L191

eiriktsarpalis replied:
Is the outer cast to `long` necessary?

lateapexearlyspeed replied:
Because we have introduced "continuously minimumGrow" logic as suggested for List, so if not set capacity to 1 for empty array in initial newCapacity, following loop will be dead loop.

eiriktsarpalis replied:
I'm not sure I understand, could you elaborate?

lateapexearlyspeed replied:
For example, if array is empty:
int newcapacity = (int)(_array.Length * (long)GrowFactor / 100);  // newcapacity will be 0
So that 
while (newCapacity < capacity) //  make loop dead
{
                    newCapacity = (int)((long)newCapacity * (long)GrowFactor / 100);  // newcapacity will be 0 forever
}

eiriktsarpalis replied:
> int newcapacity = (int)(_array.Length * (long)GrowFactor / 100); // newcapacity will be 0

Couldn't you make this like the `List` implementation?

```csharp
int newCapacity = _array.Length == 0 ? MinimumGrow : (int)(_array.Length * (long)GrowFactor / 100);
```

lateapexearlyspeed replied:
Fixed.

eiriktsarpalis commented on src/libraries/System.Collections/src/System/Collections/Generic/Queue.cs:
```diff
@@ -386,6 +386,64 @@ public void TrimExcess()
+                    newCapacity = _array.Length + MinimumGrow;
+                }
+
+                T[] newArray = new T[newCapacity];
```
This is duplicating the logic of the SetCapacity method. Why not just call the method here?

lateapexearlyspeed replied:
Because [comment](https://github.com/dotnet/runtime/pull/47149#issuecomment-767283712), there is a little difference.

eiriktsarpalis replied:
I think we should avoid duplicating the logic. While it would be nice to eliminate redundant checks as well, it's probably not as important since the dominant factor is the array resizing.

lateapexearlyspeed replied:
Fixed.

eiriktsarpalis commented on src/libraries/System.Collections/src/System/Collections/Generic/Queue.cs:
```diff
@@ -386,6 +386,64 @@ public void TrimExcess()
+                    if (newCapacity < capacity) newCapacity = capacity;
+                }
+
+                if (newCapacity < _array.Length + MinimumGrow)
```
Would remove this and factor in `MinimumGrow` in the original calculation.

lateapexearlyspeed replied:
Sorry I have not figure out how to remove this and ensure it can be more than minumumGrow, could you please give detail info ?

eiriktsarpalis replied:
See my comment in https://github.com/dotnet/runtime/pull/47149#discussion_r576265295

lateapexearlyspeed replied:
Fixed based on comment.
From my initial understand, MinimumGrow is to make sure every attempt should increase no less than it. If that is true, maybe its definition will be broken after removing the last check:

> var queue = new Queue(1);
queue.EnsureCapacity(2);

eiriktsarpalis commented on src/libraries/System.Private.CoreLib/src/System/Collections/Generic/List.cs:
```diff
@@ -391,22 +391,61 @@ public void CopyTo(T[] array, int arrayIndex)
+        /// If the current capacity of the list is less than specified <paramref name="capacity"/>,
+        /// the capacity is increased by continuously twice current capacity until it is at least the specified <paramref name="capacity"/>.
+        /// </summary>
+        /// <param name="capacity">The minimum capacity to ensure</param>
```
```suggestion
        /// <param name="capacity">The minimum capacity to ensure.</param>
```

eiriktsarpalis commented on src/libraries/System.Private.CoreLib/src/System/Collections/Generic/List.cs:
```diff
@@ -391,22 +391,61 @@ public void CopyTo(T[] array, int arrayIndex)
                 if (newCapacity < min) newCapacity = min;
-                Capacity = newCapacity;
             }
+
```
This should use the existing `Capacity` setter instead.

lateapexearlyspeed replied:
It is based on [comment](https://github.com/dotnet/runtime/pull/47149#discussion_r560239273)

eiriktsarpalis replied:
Oops. See my previous comment, would rather take one extra check over duplicating this logic. My original feedback might have led you down this path, apologies for the confusion.

lateapexearlyspeed replied:
No problem, just confirm: this will also use existing Capacity setter even though it will introduce more than one extra check.

eiriktsarpalis replied:
Yes, it is preferable to duplicating the logic.

lateapexearlyspeed replied:
Fixed.

eiriktsarpalis commented on src/libraries/System.Private.CoreLib/src/System/Collections/Generic/List.cs:
```diff
@@ -391,22 +391,61 @@ public void CopyTo(T[] array, int arrayIndex)
+        /// Increase the capacity of this list to at least the specified <paramref name="min"/> by continuously twice current capacity.
+        /// </summary>
+        /// <param name="min">The minimum capacity to ensure</param>
+        private void IncreaseCapacityTo(int min)
```
The name `IncreaseCapacityTo` seems to imply that `min` is the precise capacity we are setting it to. I would recommend renaming it to something like `EnsureCapacityCore`.

lateapexearlyspeed replied:
Fixed.

eiriktsarpalis commented on src/libraries/System.Private.CoreLib/src/System/Collections/Generic/List.cs:
```diff
@@ -391,22 +391,61 @@ public void CopyTo(T[] array, int arrayIndex)
+
+        private void EnsureCapacityWithoutUpdatingVersion(int capacity)
+        {
+            if (_items.Length < capacity)
```
Why does this need to be a separate method? I would incorporate this in `IncreaseCapacityTo`.

lateapexearlyspeed replied:
It is to avoid incremental version number twice in one Add method call (and remove check "capacity < 0" by the way).
If it is tricky, I will remove this method and reuse EnsureCapacity().

eiriktsarpalis replied:
Makes sense. In that case, couldn't the `if (_items.Length < capacity)` check incorporated in the `IncreaseCapacityTo` method? It might be redundant in certain cases but probably not that significant all things considered.

lateapexearlyspeed replied:
Fixed.

eiriktsarpalis commented on src/libraries/System.Private.CoreLib/src/System/Collections/Generic/List.cs:
```diff
@@ -391,22 +391,61 @@ public void CopyTo(T[] array, int arrayIndex)
+        private void IncreaseCapacityTo(int min)
+        {
+            int newCapacity = _items.Length == 0 ? DefaultCapacity : _items.Length * 2;
+            while (newCapacity < min)
```
This logic doesn't account for integer overflow.

lateapexearlyspeed replied:
Fixed.

eiriktsarpalis commented on src/libraries/System.Collections/src/System/Collections/Generic/Queue.cs:
```diff
@@ -386,6 +386,64 @@ public void TrimExcess()
+            if (_array.Length < capacity)
+            {
+                int newCapacity = (int)((long)(_array.Length == 0 ? 1 : _array.Length) * (long)GrowFactor / 100);
+                while (newCapacity < capacity)
```
Similar to the list implementation, this should account for arithmetic overflow.

lateapexearlyspeed replied:
Fixed: because GrowFactor is a config (not always 200) so here uses another way to handle overflow.

eiriktsarpalis commented on src/libraries/System.Collections/src/System/Collections/Generic/Stack.cs:
```diff
@@ -288,6 +288,45 @@ private void PushWithResize(T item)
+            if (_array.Length < capacity)
+            {
+                int newCapacity = (_array.Length == 0) ? DefaultCapacity : 2 * _array.Length;
+                while (newCapacity < capacity)
```
Same here, account for arithmetic overflow.

lateapexearlyspeed replied:
Fixed.

lateapexearlyspeed committed:
issue-44801 Fix comments: consider integer overflow; refactor methods.

lateapexearlyspeed committed:
Update src/libraries/System.Private.CoreLib/src/System/Collections/Generic/List.cs

Fix comment: update xml doc.

Co-authored-by: Eirik Tsarpalis <eirik.tsarpalis@gmail.com>

eiriktsarpalis committed:
ensure same algorithm is used for all resize operations

eiriktsarpalis commented:
Hey @lateapexearlyspeed, I just pushed a commit that makes a few adjustments to your PR:

1. Ensure there is only one resize algorithm defined in each class.
2. I reverted the `while (newcapacity < capacity) newcapacity *= 2;` logic, since it would result in different allocation patterns in methods such as `List.AddRange`.
3. Added tests validating `EnsureCapacity` behaviour when the argument is larger than the maximum array size.
4. Fixed a bug in `Stack<T>` that would throw `ArgumentOutOfRangeException` instead of `OutOfMemoryException` when attempting to push more elements than the maximum array size permits.

Hope it makes sense.

eiriktsarpalis committed:
revert array length check in List.EnsureCapacityCore

lateapexearlyspeed commented:
Hi @eiriktsarpalis I have no question about change, thanks. 
Please note new test "EnsureCapacity_LargeCapacity_Throws" does not throw exception.

And would confirm one thing about following code in Queue.cs about overflow:

>             int newcapacity = (int)((long)_array.Length * GrowFactor / 100);
>             // Ensure minimum growth and account for arithmetic overflow.
>             if (newcapacity < _array.Length + MinimumGrow) newcapacity = _array.Length + MinimumGrow;
In future, if GrowFactor is not 200 (for example 400), then it is possible that newcapacity is indeed overflowed after first line calculation but it skipped third line check because int newcapacity value may be larger than original newcapacity after multiple 4, for example take a 8-bit signed number: (0101 1000) << 2 = (0110 0000).
So we can accept that case as long as the calculation can finish and get value which meets MinimumGrow and requested capacity, right ?

eiriktsarpalis committed:
remove GrowFactor constant

eiriktsarpalis commented:
> Please note new test "EnsureCapacity_LargeCapacity_Throws" does not throw exception.

Could you elaborate?

> In future, if GrowFactor is not 200 (for example 400), then it is possible that newcapacity is indeed overflowed after first line calculation but it skipped third line check because int newcapacity value may be larger than original newcapacity after multiple 4, for example take a 8-bit signed number: (0101 1000) << 2 = (0110 0000).
So we can accept that case as long as the calculation can finish and get value which meets MinimumGrow and requested capacity, right ?

`GrowFactor` is a constant we are unlikely to ever change. I stopped short of removing it altogether in my changes but it is essentially the same algorithm as the other two types (modulo the `MinimumGrow` guarantee).

lateapexearlyspeed commented:
> > Please note new test "EnsureCapacity_LargeCapacity_Throws" does not throw exception.
> 
> Could you elaborate?

No issue, just notice that new test failed :)

lateapexearlyspeed commented:
> 2\. I reverted the `while (newcapacity < capacity) newcapacity *= 2;` logic, since it would result in different allocation patterns in methods such as `List.AddRange`.

BTW could you help explain "different allocation pattern" ? I know it will definitely have chance to give different newcapacity result if using  `while (newcapacity < capacity) newcapacity *= 2;`.

eiriktsarpalis commented:
> BTW could you help explain "different allocation pattern" ? I know it will definitely have chance to give different newcapacity result if using while (newcapacity < capacity) newcapacity *= 2;.

Calling `new List<int>().AddRange(Enumerable.Range(1, 70))` allocates a buffer of size 70. OTOH calling `new List<int>().EnsureCapacity(70)` would allocate a buffer of size 128. The two should be consistent, however if we changed `AddRange` to use the same logic it would result in nontrivial increase of memory consumption.

> No issue, just notice that new test failed :)

Oh. I guess it worked on my machine 😄

eiriktsarpalis commented:
>> No issue, just notice that new test failed :)
>
> Oh. I guess it worked on my machine 😄

Seems like a mono issue. I just checked and the following is valid in mono 6.12:
```csharp
var arr = new int[int.MaxValue];
arr[int.MaxValue - 1] = 42;
```
So we have two options: either remove the tests or apply the `SkipOnMono` attribute. I'm wondering though if there's a way to only run a given test in CoreCLR. I couldn't find any attribute magic to enable such behaviour. @ViktorHofer might know if it's possible.

eiriktsarpalis committed:
skip large capacity tests on mono.

eiriktsarpalis committed:
Fix overflow handling when capacity < MaxArrayLength; add clarifying comments

ViktorHofer commented:
You can either use the `SkipOnMonoAttribute` or use the `ActiveIssueAttribute`, more precisely this ctor: https://github.com/dotnet/arcade/blob/1ca4a4637bf32494c8ba579c32e4512309cae751/src/Microsoft.DotNet.XUnitExtensions/src/Attributes/ActiveIssueAttribute.cs#L21

eiriktsarpalis commented:
Ok, so if I wanted to have a test guaranteed to only run in CoreCLR (assuming we eventually add more entries to the `TestRuntimes` enum in the future), I would have to write something like `[ActiveIssue("Only specifies CoreCLR behaviour", ~TestRuntimes.CoreCLR)]`?

ViktorHofer commented:
Right, but we usually only use the `ActiveIssueAttribute` in cases where a behavior needs to be fixed on the excluded platform/runtime/etc. In this case it seems there's no other option so that should be fine

eiriktsarpalis commented:
Thank you for your contribution!
